include/material/norsand.tcc

```diff
@@ -0,0 +1,589 @@
+                                                          p_image_initial_))},
+      // p_cohesion
+      {"p_cohesion", p_cohesion_initial_},
+      // p_cohesion degradation function
+      {"zeta", 1.},
+      // Equivalent plastic deviatoric strain
+      {"epds", 0.},
+      // Plastic strain components
+      {"plastic_strain0", 0.},
+      {"plastic_strain1", 0.},
+      {"plastic_strain2", 0.},
+      {"plastic_strain3", 0.},
+      {"plastic_strain4", 0.},
+      {"plastic_strain5", 0.}};
+
+  return state_vars;
+}
+
+//! Compute elastic tensor
+template <unsigned Tdim>
+bool mpm::NorSand<Tdim>::compute_elastic_tensor() {
+  // Shear modulus
+  const double G = shear_modulus_;
+  const double a1 = bulk_modulus_ + (4.0 / 3.0) * G;
+  const double a2 = bulk_modulus_ - (2.0 / 3.0) * G;
+  // compute elastic stiffness matrix
+  // clang-format off
+  de_(0,0)=a1;    de_(0,1)=a2;    de_(0,2)=a2;    de_(0,3)=0;    de_(0,4)=0;    de_(0,5)=0;
+  de_(1,0)=a2;    de_(1,1)=a1;    de_(1,2)=a2;    de_(1,3)=0;    de_(1,4)=0;    de_(1,5)=0;
+  de_(2,0)=a2;    de_(2,1)=a2;    de_(2,2)=a1;    de_(2,3)=0;    de_(2,4)=0;    de_(2,5)=0;
+  de_(3,0)= 0;    de_(3,1)= 0;    de_(3,2)= 0;    de_(3,3)=G;    de_(3,4)=0;    de_(3,5)=0;
+  de_(4,0)= 0;    de_(4,1)= 0;    de_(4,2)= 0;    de_(4,3)=0;    de_(4,4)=G;    de_(4,5)=0;
+  de_(5,0)= 0;    de_(5,1)= 0;    de_(5,2)= 0;    de_(5,3)=0;    de_(5,4)=0;    de_(5,5)=G;
+  // clang-format on
+
+  return true;
+}
+
+//! Compute stress invariants
+template <unsigned Tdim>
+bool mpm::NorSand<Tdim>::compute_stress_invariants(const Vector6d& stress,
+                                                   mpm::dense_map* state_vars) {
+
+  // Note that in this subroutine, stress is compression positive
+
+  // Compute mean stress p
+  double mean_p = (stress(0) + stress(1) + stress(2)) / 3.;
+  if (mean_p < 1.0E-15) mean_p = 1.0E-15;
+  (*state_vars).at("p") = mean_p;
+
+  // Compute the deviatoric stress
+  Vector6d dev_stress = Vector6d::Zero();
+  dev_stress(0) = stress(0) - mean_p;
+  dev_stress(1) = stress(1) - mean_p;
+  dev_stress(2) = stress(2) - mean_p;
+  dev_stress(3) = stress(3);
+  if (Tdim == 3) {
+    dev_stress(4) = stress(4);
+    dev_stress(5) = stress(5);
+  }
+
+  // Compute J2
+  double j2 =
+      (pow((stress(0) - stress(1)), 2) + pow((stress(1) - stress(2)), 2) +
+       pow((stress(0) - stress(2)), 2)) /
+          6.0 +
+      pow(stress(3), 2);
+  if (Tdim == 3) j2 += pow(stress(4), 2) + pow(stress(5), 2);
+  if (fabs(j2) < 1.0E-15) j2 = 1.0E-15;
+  (*state_vars).at("j2") = j2;
+
+  // Compute J3
+  double j3 = (dev_stress(0) * dev_stress(1) * dev_stress(2)) -
+              (dev_stress(2) * pow(dev_stress(3), 2));
+  if (Tdim == 3)
+    j3 += ((2 * dev_stress(3) * dev_stress(4) * dev_stress(5)) -
+           (dev_stress(0) * pow(dev_stress(4), 2)) -
+           (dev_stress(1) * pow(dev_stress(5), 2)));
+  (*state_vars).at("j3") = j3;
+
+  // Compute q
+  double deviatoric_q = sqrt(3 * j2);
```

Comment:
It's always better to use std::sqrt() instead of sqrt().
See: https://stackoverflow.com/questions/26547095/why-sqrt-in-global-scope-is-much-slower-than-stdsqrt-in-mingw

Reply:
I will use `std::` for everything now. Not sure why I didn't use it.

Reply:
I will use it for all :) I need to use it for `pow` too!